### ISSUE TITLE
Loses Google Calendar

### ISSUE THREAD

**Issue by HankVanZile** [sfsam/Itsycal#51]
I'm experiencing a bug where Itsycal loses access to a Google calendar I've selected to be displayed. On a regular basis I look at Itsycal it's just not there anymore - events no longer show up in the list and the calendar is not available in the list of calendars to choose from in the preferences screen.

The calendar is added as an account to the Mac Calendar app and has no problems displaying there.

Restarting Itsycal takes care of the issue every time - the events all show up in the list and the calendar appears as selected in the preferences window.

**Comment by sfsam** [comment]
@HankVanZile Since late January, I've had a few people email me with this problem. I also use Google Calendar, but am not experiencing it so I'm stumped at the moment.

Have you been using Itsycal for long and only recently noticed this issue? Can you think of anything that might be triggering it? Any more context you could provide would be helpful.

I'm also wondering if you could try using older versions of the app and see if those older versions also have this problem?

**Comment by HankVanZile** [comment]
@sfsam Thanks for getting back to me. I've been using Itsycal for a while (since Oct of 2015, I believe) and it definitely didn't do this when I first started using the app. That said, it's been happening for a least a few versions now because I was recall being excited when 0.11.2 was released, hoping it would fix the problem. (It only just occurred to me to see if there was somewhere I could report a bug, sorry.)

I can't think of anything that might be triggering it.  Like I said, I'm not losing access to the account in Calendar.app or anything like that.  It seems to randomly happen every day or two and I only notice when I click on the app icon to see my list of meetings.  

I have it displaying a Microsoft Exchange calendar, too, and it *never* drops that one from the list. 

Regarding older versions of the app, sure, I'd be willing to try. I'm running macOS Sierra 10.12.3, so feel free to point me to any versions that are compatible with my OS.

Thanks!

**Comment by sfsam** [comment]
@HankVanZile Older versions can be downloaded from the version history page. I think they should all work for you since you have the most recent macOS.

https://www.mowglii.com/itsycal/versionhistory.html

**Comment by HankVanZile** [comment]
Thanks, @sfsam. I have removed 0.11.2 with AppCleaner (including all support folders, plist files, etc.), restarted and reinstalled 0.10.16. I wanted to go back a few versions but not too far. I'll monitor and check back in with an update if it drops my Google Calendar - or in a few days if it doesn't.

**Comment by sfsam** [comment]
@HankVanZile Thank you.

**Comment by HankVanZile** [comment]
@sfsam I can now verify that this bug existed in 0.10.16, as Itsycal lost my Google Calendar yesterday afternoon with the exact same symptoms reported above. 

I'm happy to keep going backward. What version should I try next?

**Comment by sfsam** [comment]
@HankVanZile Interesting. That version was released at the end of October 2016. What seems weird to me is that I only started getting notified from a few people of this issue at the end of January 2017.

Could you go as far back as version 0.10.10? That was released in February 2016.

**Comment by HankVanZile** [comment]
@sfsam I agree - I'm pretty sure I haven't been having this problem since Oct. 2016. Could it be something in your code that is no longer compatible as of OS 10.12.3 (released January 23, 2017) or 10.12.2 (released Dec. 13/14, 2016)?  It does appear that a few Calendar changes we in 10.12.2, for whatever that's worth.

I will uninstall 0.10.16 tonight and install 0.10.10.

**Comment by HankVanZile** [comment]
Hi @sfsam - it happened this afternoon with 0.10.10.

**Comment by sfsam** [comment]
@HankVanZile Thank you. This is helpful. I don't think there's much point in going back further right now. 

So... the bug doesn't seem to be related to anything recently changed in Itsycal since it happens with a version from over a year ago. Does that sound right?

I guess it seems like something recently changed with either the OS or the way the system notifies apps of calendar updates or authenticates access to certain calendars? I'm also on the latest Sierra and I still haven't had this happen. I don't think it is happening to a lot of people, but I get an email about it every so often. I just got another one a few days ago. Before Jan 29, 2017, I don't think anyone had emailed me about this.

Can you think of anything about the way your system works that might be different? Maybe something to do with authentication or networking?

**Comment by HankVanZile** [comment]
@sfsam I completely agree. Seems like it's something outside of Itsycal. The only other factor I can think of is that the Google account that provides the calendar is a Google Apps account. Could that have something to do with it?

**Comment by sfsam** [comment]
@HankVanZile I have no idea, but I'm glad you mentioned that. My Google account is just a regular personal account. I will ask the other people who have emailed me if they are also using Google Apps.

I'm not familiar with how Google Apps works (or even exactly what it is). Is there anything you can think of that makes it different from a personal Google account?

**Comment by HankVanZile** [comment]
@sfsam I have not seen this bug happen in a number of months at this point.  I'm going to go ahead and close this. I assume whatever was happening has been fixed - even if we never identified it!

**Comment by sfsam** [comment]
@HankVanZile Thank you, Hank.

**Comment by jumban** [comment]
After updating to the latest version of Itsycal (0.11.14), I have been experiencing this issue repeatedly.

**Comment by sfsam** [comment]
@jumban What version of macOS are you on? How often does it happen? Do the calendars come back if you quit Itsycal and restart it? Any color you can provide would be helpful.

I've also noticed this occasionally on macOS 10.14.4. From what I can tell, it looks like the Apple API I rely on to get calendar info sometimes returns incomplete data. I'm still looking into it.

In the meantime, I've made a test version that enables you to reset the calendar data connection by pressing `command-option-r`. It's not meant as a fix for this issue, but can function as a band-aid. You can try it out by downloading from the link below and replacing your version of Itsycal with it. If you do try it, I'd appreciate if you'd let me know if it restores you missing calendars.

https://s3.amazonaws.com/itsycal/Itsycal-reset.zip

**Comment by jumban** [comment]
> @jumban What version of macOS are you on? How often does it happen? Do the calendars come back if you quit Itsycal and restart it? Any color you can provide would be helpful.
> 
> I've also noticed this occasionally on macOS 10.14.4. From what I can tell, it looks like the Apple API I rely on to get calendar info sometimes returns incomplete data. I'm still looking into it.
> 
> In the meantime, I've made a test version that enables you to reset the calendar data connection by pressing `command-option-r`. It's not meant as a fix for this issue, but can function as a band-aid. You can try it out by downloading from the link below and replacing your version of Itsycal with it. If you do try it, I'd appreciate if you'd let me know if it restores you missing calendars.
> 
> https://s3.amazonaws.com/itsycal/Itsycal-reset.zip

Thanks for such a speedy response. I am indeed on Mojave 10.14.4. 

It happens consistently everyday, as far as I can tell. The Google calendars do come back every time I quit Itsycal and start it again. It then disappears again after a while. 

I will try out the test version soon and let you know how it goes. 

Also, thanks for such a wonderful app!

**Comment by jumban** [comment]
> @jumban What version of macOS are you on? How often does it happen? Do the calendars come back if you quit Itsycal and restart it? Any color you can provide would be helpful.
> 
> I've also noticed this occasionally on macOS 10.14.4. From what I can tell, it looks like the Apple API I rely on to get calendar info sometimes returns incomplete data. I'm still looking into it.
> 
> In the meantime, I've made a test version that enables you to reset the calendar data connection by pressing `command-option-r`. It's not meant as a fix for this issue, but can function as a band-aid. You can try it out by downloading from the link below and replacing your version of Itsycal with it. If you do try it, I'd appreciate if you'd let me know if it restores you missing calendars.
> 
> https://s3.amazonaws.com/itsycal/Itsycal-reset.zip

Hi, I have installed and used your test version. The problem does still occur and I can indeed restore the missing calendars with the `command-option-r` hotkey. 

Hope this helps you in finding a fix. Thank you!

**Comment by sfsam** [comment]
@jumban Can you give me an idea of how often during the day your Google calendars disappear? Do you have a lot of calendar activity (adding, editing, removing events)?

**Comment by jumban** [comment]
> @jumban Can you give me an idea of how often during the day your Google calendars disappear? Do you have a lot of calendar activity (adding, editing, removing events)?

I notice it happening about 2-3 times a day. I mostly use Itsycal to look at events, and I rarely add/modify my calendar events once they are created. I can confirm that the issue occurs even when I do not make any modifications to my events.

**Comment by sfsam** [comment]
@jumban I'm curious if you've upgraded to the latest version (0.11.15) and noticed any improvement?

**Comment by jumban** [comment]
> @jumban I'm curious if you've upgraded to the latest version (0.11.15) and noticed any improvement?

I have not been using my Mac much lately so I can't say for sure, but I am on the latest version now. 

I will report back again if I notice the issue happening (or not).

**Comment by jumban** [comment]
I have been using my Mac sporadically the last few weeks, and the problem has not occured at all since the last update. 

Not sure if this affects anything, but I have also moved to a different country with a different timezone since then.

**Comment by wlonkly** [comment]
I see this hasn't been updated for a bit, so wanted to chime in: this happens to my Google calendar on the regular, running 0.12.6 on Catalina. It seems to disappear toward the end of the day. Restarting restores it.

**Comment by wlonkly** [comment]
A new data point appears!

Itsycal just had my Google Calendar disappear on me, as others in this issue experienced -- and at the same time, [Meeter](https://trymeeter.com/) _also_ had the same events disappear. I thought I'd mention it because of the implication that the problem might not be inside Itsycal.
